# Pandas Data Analysis

In [1]:
import pandas as pd

- load the titanic csv data
- https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv

In [68]:
df=pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [ ]:
# dataframe and series , are two datatypes in pandas

In [5]:
#get first two records
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [6]:
df.dtypes
#column types of the dataframe

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## initial insights about the data
- what is the exact meaning of dataset? what is the exact meaning of the attribute
- never just load the data and start working, always ask the meaning of each column, it helps a lot for the datascientist,understand the data,understand the pattern and then make the model

## titanic data
- passengers are onboarded, so there are ids for passengers
- how many people have survived in the titanic accident
- name of the passenger . 

In [10]:
## df info
df.info()
### gives the column name and datatype for each column
# there is survived column, in every column we have 891 values, but in  cabin column, we have very less number of records
# meaning, if there is a dataset.and if there are less values, so there will be null values, this is the initial observation,we have to handle the null values.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
## df Describe

In [12]:
## df Describe
df.describe()
# gives the statistics, count, mean std for each column


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## analysis
- passengerid, count is 891 records
- in survived column, there is 891 column, in age there are 714,there might be some missing column,there must be some null values, we have to handle null values in data preprocessing . we have to see in  EDA . Exploratory Data Analysis. it is contextual one,it changes with dataset everytime. we have to handle the null values, we will take a note of it and try to handle it in future
### mean
- mean is the average.
- on an average, people have paid 32.
- the average age of a person is 29 .
- passenger id is not important, we cant draw any conclusion,we have to try to understand the column once.

### standard deviation
- meaning is, dispersion of dataset from mean
- what is the significance with standard deviation? when SD is high or SD is low.
- if SD is very high, then what is the effect on the top of the distribution of data, it will affect the ML model to learn the entire pattern, if SD is very less, also it will affect the model 
- Fare. average is 32, SD is 49 , from average, the fare is fluctuating a lot, example we have SD = 1. and average fare is 32. 32+-1 . buy class 1,class2 ,class 3 tickets. we would chose class 1 ? ,10-15 year back, flight tickets were not booked, bus and train tickets were booked. but now there are many people who are buying the flight tickets, the differences is, the fare margin from train ticket and flight ticket,it was huge in 10-15 years . so we would go for train,since it would cost a lot. now if we check the train fare, and flight ticket, they are approximately close .so SD is almost negligble ,or very less for a train and flight ticket, so we chose flight instead of train. so we can say that the from the average, SD of flight ticket has decreased . 

### variances
- variances => driving a bike or car. speed of the vehicle (0,10,80,100,40) ,what is the variance of the speed? => 0 to 100. what is the standard deviation of the speed? we take the average and , we will see how much from left and right side it is fluctuating. for ticket fare , it varies a lot. one month , one day,one hour before, there will be huge variations of the fares of the tickets.

### min 
- min fare of all the tickets. 0 is the ticket . min age of person who got onboarded is 0.42 .

### 25% ,50%,75%
- ex: we take a number , (1,2,4,1,8,10,11,2,8,9,10,1) . arrange the numbers in ascending order . 
- (1,1,1,2,2,4,8,8,9,10,11) , the min of the numbers is 1.maximum is 11 .now 25% percentile . 25% doesnt mean that 25 % of data, it means 25th data out of 100 data. ex: we have 12 indexes, what is the 25% percentile indexes? it is 0.25*12=3.
- 25th percentile. instead of using min and max, we can use 0th percentile and 100th percentile respectively
- it is representing indexes and not percentage of data
- if there is a chance that, it is 12 numbers, we can find 25th percentile by taking 25th percentile of the 12, if we have 13 numbers, we will take average of 3rd and 4th and then takes average and gives out the data
- for 891, it is not even to split,so it takes average of two values .

# doubts
- Standard deviation and variance is still not clear

In [15]:
df.columns
#here df.describe didnt consider all the columns,it just considered the column which is datatypes as numeric one,it didnt consider categorical, what about the columns as object datatype

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [16]:
#by default,if we are calling df.describe,it will just consider the numeric columns,it wont consider categorical ones, we
# for categorical,how will we find the min and max? we cant find min ,max, SD in between the names

In [18]:
#filter the columns which have the datatypes = objects
df.dtypes == "object"
# wherever datatype is object, it is returning false, since we gave a comparision operator
# now objective is to filter out all the columns where datatypes = objects

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [19]:
type(df.dtypes)

pandas.core.series.Series

## Generate list of columns with object datatype

In [21]:
# series is equivalent to list
## only difference is, list represents only values, whereas series represents its index name as well
df.dtypes[df.dtypes=="object"]
# give all the dtypes of columns where the dtypes=="object"

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

In [33]:
type(df.dtypes[df.dtypes=="object"])
#this is series

pandas.core.series.Series

In [22]:
l=[4,5,6,7,8,9,0,"sudh","kumar"]

In [23]:
l[0]

4

In [25]:
l[l==8] #it gives the indexes of the element 8,it gives the first occurence
#incase of series, it gives the indexes and respective datatype

4

## generates index of the filter of object datatype

In [50]:
df.dtypes=="object"

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
ineuron         True
dtype: bool

In [49]:
df.dtypes[df.dtypes=="object"]

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
ineuron     object
dtype: object

In [28]:
df.dtypes[df.dtypes=="object"].index
# we are getting list of columns wherever datatype = object
# we are able to extrac those columns where columns are object
col_name= df.dtypes[df.dtypes=="object"].index
col_name

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [34]:
type(col_name)# type is index

pandas.core.indexes.base.Index

In [30]:
df[col_name]
# now we can filter the columns, with data type = object

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

## Describing the object datatypes

In [32]:
# for numeric columns, we got the description, for categorical we didtnt get,now we can call describe on it
df[col_name].describe()
# hamming distance=> counting the number of mismatches of the characters between two strings

#here there  is 891 for name,sex. for cabin column, there are some missing columns, for embarked, there are 2 missing columns
#unique tickets are 681. but total is 891 . there might be reptition of it.
# total cabins are 204, unique are 147

## top and frequency
# frequency of top record=1
# frequency of male is 577
# frequency of ticket 347082 is 7


Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

## add a new column to this dataset

In [37]:
df['ineuron']="rohan"
#column_name, just like we add a new key to the dictoinary, it duplicates the same value across all the rows

In [36]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked ineuron  
0        0         A/5 21171   7.2500   NaN        S   rohan  
1        0          PC 17599  71.2833   C85        C   rohan  
2        0  STON/O2. 3101282   7.9250   NaN        S   rohan  
3        0            113803  53.1000  C123        S   rohan  
4        0            373450   8.0500   NaN        S   rohan  
..     ...               ...      ...   ...      ...     ...  
886      0            211536  13.0000   NaN        S   rohan  
887      0            112053  30.0000   B42        S   rohan  
888      2        W./C. 6607  23.4500   NaN        S   rohan  
889      0            111369  30.0000  C148        C   rohan  
890      0            370376   7.7500   NaN        Q   rohan  

[891 rows x 13 columns]

## Extract some of the columns

In [38]:
## Extract some of the columns
df['Name']
## extract the name column

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

## extract first 15 records

In [42]:
## extract first 15 records
df['Name'][:15:]
## series is almost equal to list

0                               Braund, Mr. Owen Harris
1     Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                Heikkinen, Miss. Laina
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                              Allen, Mr. William Henry
5                                      Moran, Mr. James
6                               McCarthy, Mr. Timothy J
7                        Palsson, Master. Gosta Leonard
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                   Nasser, Mrs. Nicholas (Adele Achem)
10                      Sandstrom, Miss. Marguerite Rut
11                             Bonnell, Miss. Elizabeth
12                       Saundercock, Mr. William Henry
13                          Andersson, Mr. Anders Johan
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

### Extract alternate records from 0 to 15

In [43]:
## Extract alternate records from 0 to 15
df['Name'][:15:2]

0                               Braund, Mr. Owen Harris
2                                Heikkinen, Miss. Laina
4                              Allen, Mr. William Henry
6                               McCarthy, Mr. Timothy J
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                      Sandstrom, Miss. Marguerite Rut
12                       Saundercock, Mr. William Henry
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

## do we have null values or not?
- in cabin , age we have null values, we want to know what row indexes, we have null values? we are able to understand that there are null  values with describe()
- we want to know the row indexes of the null values
- we can do that by extracting the age column

In [44]:
# extract the age column
df['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [46]:
# is null 
df['Age'].isnull()# returns true,false

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [47]:
# filter the true values where value is null
df['Age'].isnull() == True

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [51]:
# extract these null values indexes to a list
df[df['Age'].isnull() == True]
# gets the dataframe which has null values

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q   rohan  
17     male  NaN      0      0      244373  13.0000   NaN        S   rohan  
19   female  NaN      0      0        2649   7.2250   NaN        C   rohan  
26     male  NaN      0      0        2631   7.2250   NaN        C   rohan  
28   female  NaN      0      0      330959   7.8792   NaN        Q   rohan  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C   rohan  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S   rohan  
868    male  NaN      0      0      345777   9.5000   NaN        S   rohan  
878    male  NaN      0      0      349217   7.8958   NaN        S   rohan  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S   rohan  

[177 rows x 13 columns]

In [52]:
## get the indexes where there are null values in age column
df[df['Age'].isnull() == True].index

Int64Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
            ...
            832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
           dtype='int64', length=177)

In [54]:
len(df[df['Age'].isnull() == True].index)

177

## Show all the record values where we have null values for the indexes
- how to get the dataset from the list of indexes. we need row wise selection

In [58]:
indexes= df[df['Age'].isnull() == True].index
indexes

Int64Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
            ...
            832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
           dtype='int64', length=177)

### Filter using with using Loc, iloc

In [60]:
df.loc[indexes]
#now we are able to see all the records where Age = null

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q   rohan  
17     male  NaN      0      0      244373  13.0000   NaN        S   rohan  
19   female  NaN      0      0        2649   7.2250   NaN        C   rohan  
26     male  NaN      0      0        2631   7.2250   NaN        C   rohan  
28   female  NaN      0      0      330959   7.8792   NaN        Q   rohan  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C   rohan  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S   rohan  
868    male  NaN      0      0      345777   9.5000   NaN        S   rohan  
878    male  NaN      0      0      349217   7.8958   NaN        S   rohan  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S   rohan  

[177 rows x 13 columns]

In [63]:
df["Age"].isnull()==True
# gives the series of indexes with true or false values

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888     True
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [62]:
type(df["Age"].isnull()==True)

pandas.core.series.Series

In [61]:
### without using loc,iloc

df[df["Age"].isnull()==True]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q   rohan  
17     male  NaN      0      0      244373  13.0000   NaN        S   rohan  
19   female  NaN      0      0        2649   7.2250   NaN        C   rohan  
26     male  NaN      0      0        2631   7.2250   NaN        C   rohan  
28   female  NaN      0      0      330959   7.8792   NaN        Q   rohan  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C   rohan  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S   rohan  
868    male  NaN      0      0      345777   9.5000   NaN        S   rohan  
878    male  NaN      0      0      349217   7.8958   NaN        S   rohan  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S   rohan  

[177 rows x 13 columns]

## Find the record who has paid the highest Fare

In [73]:
df["Fare"]

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: Fare, Length: 891, dtype: float64

In [67]:
## find the record who has paid the highest fare
max(df["Fare"]) # max fare

512.3292

In [74]:
#filter out the fare with max(fare)
df[df["Fare"]==max(df["Fare"])]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

    ineuron  
258   rohan  
679   rohan  
737   rohan

In [75]:
### find the person name who has paid the highest fare
df[df["Fare"]==max(df["Fare"])]["Name"]

258                      Ward, Miss. Anna
679    Cardeza, Mr. Thomas Drake Martinez
737                Lesurer, Mr. Gustave J
Name: Name, dtype: object

# Task
1. find out how many male passengers and female passengers are onbaorded 
2. how many survived we have .
3. how many casuality we have
4. what is name of a person who is the eldest one.
5. how many passengers  do we have in first,second,third class 
6. how many person we have whose name starts with "s"
7. try to create a new column which is summation of "SibSp" and "parch"
8. how many person do we have below age of 25
9. how many person died whose age was less than 40
10. from a cabin colum ,seperate text and numeric value

### 1.find out how many male passengers and female passengers are onbaorded

In [90]:
# 1.find out how many male passengers and female passengers are onbaorded
male= len(df[df['Sex']=='male'])
print(male) #male count
# female = len( df[df['Sex']=='Female'])
# print(male,female)

577


In [91]:
female = len(df[df['Sex']=='female'])
print(female) # female count

314


#### one liner -> value counts

In [69]:
df['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [71]:
df['Sex'].value_counts()['male']

577

In [72]:
df['Sex'].value_counts()['female']

314

### 2.how many survived we have

In [97]:
survived = len(df[df['Survived']==1])
surviveda

342

In [74]:
df['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

#### one liner shortcut - value_counts

In [76]:
df['Survived'].value_counts()[0]

549

In [78]:
df['Survived'].value_counts()[1]

342

### 3.how many casuality we have

In [99]:
casuality = len(df[df['Survived']==0])
casuality

549

### 4.what is name of a person who is the eldest one.

In [102]:
df[df['Age']==max(df['Age'])]['Name']

630    Barkworth, Mr. Algernon Henry Wilson
Name: Name, dtype: object

In [82]:
df[df['Age']==df['Age'].max()]

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch Ticket  Fare Cabin Embarked  
630  male  80.0      0      0  27042  30.0   A23        S

In [83]:
df[df['Age']==df['Age'].max()]['Name']

630    Barkworth, Mr. Algernon Henry Wilson
Name: Name, dtype: object

### 5.how many passengers do we have in first,second,third class

In [106]:
classOne = len(df[df['Pclass']==1])
classTwo = len(df[df['Pclass']==2])
classThree = len(df[df['Pclass']==3])
print(classOne,classTwo,classThree)

216 184 491


#### one liner solution -> value counts

In [84]:
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

### 6.how many person we have whose name starts with "s"

In [144]:
# df[df['Name'][0] == 's']
df['Name']

df[df['Name'].str[0]=='S']

PassengerId  Survived  Pclass  \
10            11         1       3   
12            13         0       3   
23            24         1       1   
31            32         1       1   
48            49         0       3   
..           ...       ...     ...   
851          852         0       3   
862          863         1       1   
863          864         0       3   
880          881         1       2   
884          885         0       3   

                                                  Name     Sex   Age  SibSp  \
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
12                      Saundercock, Mr. William Henry    male  20.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
48                                 Samaan, Mr. Youssef    male   NaN      2   
..                                                 ...     ...   ...    ...   
851                                Svensson, Mr. Johan    male  74.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
863                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN      8   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   

     Parch           Ticket      Fare Cabin Embarked  sum_sibsp_parch  
10       1          PP 9549   16.7000    G6        S              806  
12       0        A/5. 2151    8.0500   NaN        S              806  
23       0           113788   35.5000    A6        S              806  
31       0         PC 17569  146.5208   B78        C              806  
48       0             2662   21.6792   NaN        C              806  
..     ...              ...       ...   ...      ...              ...  
851      0           347060    7.7750   NaN        S              806  
862      0            17466   25.9292   D17        S              806  
863      2         CA. 2343   69.5500   NaN        S              806  
880      1           230433   26.0000   NaN        S              806  
884      0  SOTON/OQ 392076    7.0500   NaN        S              806  

[86 rows x 13 columns]

In [85]:
len(df[df['Name'].str[0]=='S'])

86

In [89]:
df[df['Name'].str.startswith('S')]

PassengerId  Survived  Pclass  \
10            11         1       3   
12            13         0       3   
23            24         1       1   
31            32         1       1   
48            49         0       3   
..           ...       ...     ...   
851          852         0       3   
862          863         1       1   
863          864         0       3   
880          881         1       2   
884          885         0       3   

                                                  Name     Sex   Age  SibSp  \
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
12                      Saundercock, Mr. William Henry    male  20.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
48                                 Samaan, Mr. Youssef    male   NaN      2   
..                                                 ...     ...   ...    ...   
851                                Svensson, Mr. Johan    male  74.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
863                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN      8   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   

     Parch           Ticket      Fare Cabin Embarked  
10       1          PP 9549   16.7000    G6        S  
12       0        A/5. 2151    8.0500   NaN        S  
23       0           113788   35.5000    A6        S  
31       0         PC 17569  146.5208   B78        C  
48       0             2662   21.6792   NaN        C  
..     ...              ...       ...   ...      ...  
851      0           347060    7.7750   NaN        S  
862      0            17466   25.9292   D17        S  
863      2         CA. 2343   69.5500   NaN        S  
880      1           230433   26.0000   NaN        S  
884      0  SOTON/OQ 392076    7.0500   NaN        S  

[86 rows x 12 columns]

### 7. try to create a new column which is summation of "SibSp" and "Parch"

In [119]:
sum_sibsp_parch = sum(df['SibSp']) + sum(df['Parch'])
df['sum_sibsp_parch']=sum_sibsp_parch

In [120]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  sum_sibsp_parch  
0        0         A/5 21171   7.2500   NaN        S              806  
1        0          PC 17599  71.2833   C85        C              806  
2        0  STON/O2. 3101282   7.9250   NaN        S              806  
3        0            113803  53.1000  C123        S              806  
4        0            373450   8.0500   NaN        S              806  
..     ...               ...      ...   ...      ...              ...  
886      0            211536  13.0000   NaN        S              806  
887      0            112053  30.0000   B42        S              806  
888      2        W./C. 6607  23.4500   NaN        S              806  
889      0            111369  30.0000  C148        C              806  
890      0            370376   7.7500   NaN        Q              806  

[891 rows x 13 columns]

In [90]:
df['SibSp'].sum()

466

In [91]:
df['Parch'].sum()

340

In [93]:
df['new_sum_sibsip_parch']=df['SibSp']+df['Parch']
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_sum_sibsip_parch  
0        0         A/5 21171   7.2500   NaN        S                     1  
1        0          PC 17599  71.2833   C85        C                     1  
2        0  STON/O2. 3101282   7.9250   NaN        S                     0  
3        0            113803  53.1000  C123        S                     1  
4        0            373450   8.0500   NaN        S                     0  
..     ...               ...      ...   ...      ...                   ...  
886      0            211536  13.0000   NaN        S                     0  
887      0            112053  30.0000   B42        S                     0  
888      2        W./C. 6607  23.4500   NaN        S                     3  
889      0            111369  30.0000  C148        C                     0  
890      0            370376   7.7500   NaN        Q                     0  

[891 rows x 13 columns]

### 8.how many person do we have below age of 25

In [123]:
len(df[df['Age']<25])

278

### 9.how many person died whose age was less than 40

In [167]:
# len(df[df['Age']<40 and df['Survived']==0])
# df['Survived']==0 
len(df[ (df['Age']<40 )& (df['Survived']==0) ])

322

### 10.from a cabin colum ,seperate text and numeric value

In [158]:
def seperateTextAndNumber(string):
    text=""
    numeric=""
    for i in string:
#         if type(i)
#         print(type(i))
        if i >='0' and i <='9':
            numeric+=i
        elif i == ' ':
            pass
        else:
            text+=i
            
    return text,numeric

In [172]:
seperateTextAndNumber("C23 C25 C27")

('CCC', '232527')

In [22]:
import pandas as pd

# Create a sample DataFrame with a mixed column
df = pd.DataFrame({
    "col1": ["apple", 1, "banana", 2, "cherry", 3.5]
})

# Extract text values
text_values = df[df["col1"].astype(str).str.isalpha()]["col1"]
print("Text values:")
print(text_values)

# Extract numeric values
numeric_values = df[df["col1"].astype(str).str.isnumeric()]["col1"]
print("Numeric values:")
print(numeric_values)



Text values:
0     apple
2    banana
4    cherry
Name: col1, dtype: object
Numeric values:
1    1
3    2
Name: col1, dtype: object


In [23]:
df['col1']

0     apple
1         1
2    banana
3         2
4    cherry
5       3.5
Name: col1, dtype: object

In [26]:
df['col1'].astype(str).str.isalpha()

0     True
1    False
2     True
3    False
4     True
5    False
Name: col1, dtype: bool

In [173]:
# df['Cabin']
import math
text=[]
numeric=[]
for i in df['Cabin']:
    if type(i)!=str:
#         print(math.isnan(i))
#         pass
        text.append('nan')
        numeric.append('nan')
    else:
#         print(i)
        current_text,current_numeric=seperateTextAndNumber(i)
#         print(current_text,current_numeric)
        text.append(current_text)
        numeric.append(current_numeric)
        
#         print(type(i))
    

In [163]:
text

['nan',
 'C',
 'nan',
 'C',
 'nan',
 'nan',
 'E',
 'nan',
 'nan',
 'nan',
 'G',
 'C',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'D',
 'nan',
 'A',
 'nan',
 'nan',
 'nan',
 'CCC',
 'nan',
 'nan',
 'nan',
 'B',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'D',
 'nan',
 'B',
 'C',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'B',
 'C',
 'nan',
 'nan',
 'nan',
 'F',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'FG',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'CCC',
 'nan',
 'nan',
 'nan',
 'E',
 'nan',
 'nan',
 'nan',
 'A',
 'DD',
 'nan',
 'nan',
 'nan',
 'nan',
 'D',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'C',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'BB',
 'nan',
 'nan',
 'nan',
 'nan',
 'E',
 'D',
 'nan',
 'nan',
 'nan',
 'FE',
 'nan',
 'na

In [164]:
df['text']=text
df['numeric']=numeric
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  sum_sibsp_parch text  \
0        0         A/5 21171   7.2500   NaN        S              806  nan   
1        0          PC 17599  71.2833   C85        C              806    C   
2        0  STON/O2. 3101282   7.9250   NaN        S              806  nan   
3        0            113803  53.1000  C123        S              806    C   
4        0            373450   8.0500   NaN        S              806  nan   
..     ...               ...      ...   ...      ...              ...  ...   
886      0            211536  13.0000   NaN        S              806  nan   
887      0            112053  30.0000   B42        S              806    B   
888      2        W./C. 6607  23.4500   NaN        S              806  nan   
889      0            111369  30.0000  C148        C              806    C   
890      0            370376   7.7500   NaN        Q              806  nan   

    numeric  
0       nan  
1        85  
2       nan  
3       123  
4       nan  
..      ...  
886     nan  
887      42  
888     nan  
889     148  
890     nan  

[891 rows x 15 columns]

#### applying regex on  the strings

In [95]:
import re

# regular expressions is used for pattern matching or string matching

replace = re.compile("([a-zA-Z]+)")

df['string_cabin_list']=df['Cabin'].str.extract(replace)
# we are extracting the strings with replace pattern
df['numeric_cabin_list']=df['Cabin'].str.replace(replace,"")
# we are replacing the replace pattern with ""

In [96]:
df['string_cabin_list']

0      NaN
1        C
2      NaN
3        C
4      NaN
      ... 
886    NaN
887      B
888    NaN
889      C
890    NaN
Name: string_cabin_list, Length: 891, dtype: object

In [97]:
df['numeric_cabin_list']

0      NaN
1       85
2      NaN
3      123
4      NaN
      ... 
886    NaN
887     42
888    NaN
889    148
890    NaN
Name: numeric_cabin_list, Length: 891, dtype: object

In [98]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_sum_sibsip_parch  \
0        0         A/5 21171   7.2500   NaN        S                     1   
1        0          PC 17599  71.2833   C85        C                     1   
2        0  STON/O2. 3101282   7.9250   NaN        S                     0   
3        0            113803  53.1000  C123        S                     1   
4        0            373450   8.0500   NaN        S                     0   
..     ...               ...      ...   ...      ...                   ...   
886      0            211536  13.0000   NaN        S                     0   
887      0            112053  30.0000   B42        S                     0   
888      2        W./C. 6607  23.4500   NaN        S                     3   
889      0            111369  30.0000  C148        C                     0   
890      0            370376   7.7500   NaN        Q                     0   

    string_cabin_list numeric_cabin_list  
0                 NaN                NaN  
1                   C                 85  
2                 NaN                NaN  
3                   C                123  
4                 NaN                NaN  
..                ...                ...  
886               NaN                NaN  
887                 B                 42  
888               NaN                NaN  
889                 C                148  
890               NaN                NaN  

[891 rows x 15 columns]

# Load Bank Dataset
- https://archive.ics.uci.edu/ml/machine-learning-databases/00222/

In [100]:
df=pd.read_csv(r'D:\coding\github projects\full_stack_datascience_ineuron\7_pandas_and_numpy\bank\bank.csv',delimiter=';')
df

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [ ]:
# if age is low, and expenses are high, then that person might take a credit card
# if age is high, and expenses are high, he is just spending what he is earning


## rules to become a good datascientist
-  understanding of each column attribute is very important, dont write any single line of code without understanding the functional aspect of EDA 
- and understanding the environmental aspect of attribute, and then write the code
- interviewers will ask about dataset,column names, it is important

1. how many campaign available in this dataset 
2. how many users do we have with housing and personal loan.
3. how many persons do we have whose age is 60+ 
4. in which month, we have triggered most of the customer
5. which mode of call is giving you more result
6. how many entrepreneurs do we have in this list give a name? 
<!-- prepare a seperate product for them , and we can sell it -->
7. how many customers do we have with negative balance. 
<!-- here we have balance column, if their credit score is negative,we wont give  a credit card -->
8. prepare a group of data based on education level 

### 1. how many campaign available in this dataset 

In [101]:
df['campaign']

0        1
1        1
2        1
3        4
4        1
        ..
4516     5
4517     1
4518    11
4519     4
4520     2
Name: campaign, Length: 4521, dtype: int64

#### using set 

In [177]:
campaigns_set=set()

for i in df['campaign']:
    campaigns_set.add(i)

In [179]:
campaigns_set
len(campaigns_set)

32

In [110]:
set(df['campaign'])

{1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 28,
 29,
 30,
 31,
 32,
 44,
 50}

In [111]:
len(set(df['campaign']))

32

#### using value counts

In [104]:
df['campaign'].value_counts()

1     1734
2     1264
3      558
4      325
5      167
6      155
7       75
8       56
9       30
10      27
11      22
12      21
13      17
14      10
15       9
16       8
17       7
18       7
25       4
19       3
20       3
28       3
24       3
32       2
21       2
23       2
22       2
31       1
29       1
50       1
30       1
44       1
Name: campaign, dtype: int64

In [105]:
type(df['campaign'].value_counts())

pandas.core.series.Series

In [106]:
len(df['campaign'].value_counts())

32

#### using unique

In [108]:
df['campaign'].unique()

array([ 1,  4,  2,  5,  3,  6, 18, 10,  9,  7, 12, 14, 13, 24, 11,  8, 29,
       32, 16, 22, 15, 30, 25, 21, 17, 19, 23, 20, 50, 28, 31, 44],
      dtype=int64)

In [109]:
len(df['campaign'].unique())

32

### 2. how many users do we have with housing and personal loan.

In [123]:
df[(df['housing']=='yes') & (df['loan']=='yes')]

age            job  marital  education default  balance housing loan  \
1      33       services  married  secondary      no     4789     yes  yes   
3      30     management  married   tertiary      no     1476     yes  yes   
9      43       services  married    primary      no      -88     yes  yes   
14     31    blue-collar  married  secondary      no      360     yes  yes   
48     32   entrepreneur   single    primary     yes     -849     yes  yes   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4502   42       services  married  secondary      no      -91     yes  yes   
4504   42    blue-collar   single  secondary      no     1080     yes  yes   
4508   42         admin.  married    unknown      no      642     yes  yes   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome    y  
1     cellular   11   may       220         1    339         4  failure   no  
3      unknown    3   jun       199         4     -1         0  unknown   no  
9     cellular   17   apr       313         1    147         2  failure   no  
14    cellular   29   jan        89         1    241         1  failure   no  
48    cellular    4   feb       204         1     -1         0  unknown   no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ...  
4502  cellular    5   feb        43         1     -1         0  unknown   no  
4504  cellular   13   may       951         3    370         4  failure  yes  
4508   unknown   16   may       509         2     -1         0  unknown   no  
4517   unknown    9   may       153         1     -1         0  unknown   no  
4520  cellular    3   apr       345         2    249         7    other   no  

[406 rows x 17 columns]

In [183]:
df[(df['housing']=='yes') & (df['loan']=='yes')]
len(df[(df['housing']=='yes') & (df['loan']=='yes')])

406

## 3. how many persons do we have whose age is 60+ 

In [189]:
len(df[df['age']>60])

127

### 4. in which month, we have triggered most of the customer

In [197]:
contacts_month = {}
max_contacts_count=0
max_contacts_month=''
for i in df['month']:
    if i not in contacts_month:
        contacts_month[str(i)]=0
    else:
        contacts_month[str(i)]+=1
        if(max_contacts_count < contacts_month[str(i)]):
            max_contacts_month=str(i)
            max_contacts_count = contacts_month[str(i)]
            
print(max_contacts_count,max_contacts_month)

    

1397 may


In [198]:
contacts_month

{'oct': 79,
 'may': 1397,
 'apr': 292,
 'jun': 530,
 'feb': 221,
 'aug': 632,
 'jan': 147,
 'jul': 705,
 'nov': 388,
 'sep': 51,
 'mar': 48,
 'dec': 19}

#### using value_counts

In [112]:
df['month'].value_counts()

may    1398
jul     706
aug     633
jun     531
nov     389
apr     293
feb     222
jan     148
oct      80
sep      52
mar      49
dec      20
Name: month, dtype: int64

In [113]:
df['month'].value_counts().max()

1398

In [118]:
df['month'].value_counts().index[0]

'may'

### 5. which mode of call is giving you more result

In [201]:
call_mode={}
max_count=0
max_count_call_mode=''
for i in df['contact']:
    key=str(i)
    if key not in call_mode:
        call_mode[key]=1
    else:
        call_mode[key]+=1
        if max_count < call_mode[key]:
            max_count = call_mode[key]
            max_count_call_mode=key
            
call_mode
max_count,max_count_call_mode

(2896, 'cellular')

In [33]:
df

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

#### using value counts

In [128]:
df_y = df[df['y']=='yes']
df_y

age            job   marital  education default  balance housing loan  \
13     20        student    single  secondary      no      502      no   no   
30     68        retired  divorced  secondary      no     4189      no   no   
33     32     management    single   tertiary      no     2536     yes   no   
34     49     technician   married   tertiary      no     1235      no   no   
36     78        retired  divorced    primary      no      229      no   no   
...   ...            ...       ...        ...     ...      ...     ...  ...   
4494   26     technician    single  secondary      no      668     yes   no   
4503   60  self-employed   married    primary      no      362      no  yes   
4504   42    blue-collar    single  secondary      no     1080     yes  yes   
4505   32         admin.    single  secondary      no      620     yes   no   
4511   46    blue-collar   married  secondary      no      668     yes   no   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
13     cellular   30   apr       261         1     -1         0  unknown  yes  
30    telephone   14   jul       897         2     -1         0  unknown  yes  
33     cellular   26   aug       958         6     -1         0  unknown  yes  
34     cellular   13   aug       354         3     -1         0  unknown  yes  
36    telephone   22   oct        97         1     -1         0  unknown  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4494    unknown   28   may       576         3     -1         0  unknown  yes  
4503   cellular   29   jul       816         6     -1         0  unknown  yes  
4504   cellular   13   may       951         3    370         4  failure  yes  
4505    unknown   26   may      1234         3     -1         0  unknown  yes  
4511    unknown   15   may      1263         2     -1         0  unknown  yes  

[521 rows x 17 columns]

In [130]:
df_y['contact'].value_counts().index[0]

'cellular'

#### using group by

In [137]:
df_groupby_cy=df.groupby(['contact','y']).count()
df_groupby_cy

age   job  marital  education  default  balance  housing  \
contact   y                                                                
cellular  no   2480  2480     2480       2480     2480     2480     2480   
          yes   416   416      416        416      416      416      416   
telephone no    257   257      257        257      257      257      257   
          yes    44    44       44         44       44       44       44   
unknown   no   1263  1263     1263       1263     1263     1263     1263   
          yes    61    61       61         61       61       61       61   

               loan   day  month  duration  campaign  pdays  previous  \
contact   y                                                             
cellular  no   2480  2480   2480      2480      2480   2480      2480   
          yes   416   416    416       416       416    416       416   
telephone no    257   257    257       257       257    257       257   
          yes    44    44     44        44        44     44        44   
unknown   no   1263  1263   1263      1263      1263   1263      1263   
          yes    61    61     61        61        61     61        61   

               poutcome  
contact   y              
cellular  no       2480  
          yes       416  
telephone no        257  
          yes        44  
unknown   no       1263  
          yes        61

In [143]:
df_groupby_cy=df.groupby(['contact','y']).count()
df_groupby_cy.loc['cellular']#we use loc for rows, 

age   job  marital  education  default  balance  housing  loan   day  \
y                                                                            
no   2480  2480     2480       2480     2480     2480     2480  2480  2480   
yes   416   416      416        416      416      416      416   416   416   

     month  duration  campaign  pdays  previous  poutcome  
y                                                          
no    2480      2480      2480   2480      2480      2480  
yes    416       416       416    416       416       416

In [145]:
df_groupby_cy.loc['cellular','yes']

age          416
job          416
marital      416
education    416
default      416
balance      416
housing      416
loan         416
day          416
month        416
duration     416
campaign     416
pdays        416
previous     416
poutcome     416
Name: (cellular, yes), dtype: int64

In [148]:
df_groupby_cy.loc['cellular','yes'][0]

416

In [149]:
df_groupby_cy.loc['telephone','yes'][0]

44

In [150]:
df_groupby_cy.loc['unknown','yes'][0]

61

In [151]:
max(df_groupby_cy.loc['cellular','yes'][0],df_groupby_cy.loc['telephone','yes'][0],df_groupby_cy.loc['unknown','yes'][0])

416

### 6. how many entrepreneurs do we have in this list give a name? 

In [204]:
len(df[df['job']=='entrepreneur'])

168

### 7. how many customers do we have with negative balance. 

In [7]:
len(df[df['balance']<0])

366

### 8. prepare a group of data based on education level 

In [11]:
education_levels={}
for i in df['education']:
    if i not in education_levels:
        education_levels[str(i)]=1
    else:
        education_levels[str(i)]+=1
        
education_levels
#     print(i)

{'primary': 678, 'secondary': 2306, 'tertiary': 1350, 'unknown': 187}

In [18]:
primaryDf=df[df['education']=='primary']
secondaryDf= df[df['education']=='secondary']
tertiaryDf= df[df['education']=='tertiary']
unknownDf= df[df['education']=='unknown']

In [16]:
primaryDf

age            job   marital education default  balance housing loan  \
0      30     unemployed   married   primary      no     1787      no   no   
9      43       services   married   primary      no      -88     yes  yes   
18     25    blue-collar    single   primary      no     -221     yes   no   
26     55    blue-collar   married   primary      no      627     yes   no   
36     78        retired  divorced   primary      no      229      no   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4480   23    blue-collar   married   primary      no     1158     yes   no   
4485   53    blue-collar   married   primary      no      238     yes   no   
4486   37    blue-collar   married   primary      no      378     yes   no   
4499   45    blue-collar  divorced   primary      no      942      no   no   
4503   60  self-employed   married   primary      no      362      no  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
0      cellular   19   oct        79         1     -1         0  unknown   no  
9      cellular   17   apr       313         1    147         2  failure   no  
18      unknown   23   may       250         1     -1         0  unknown   no  
26      unknown    5   may       247         1     -1         0  unknown   no  
36    telephone   22   oct        97         1     -1         0  unknown  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4480   cellular   16   apr       743         1     -1         0  unknown   no  
4485   cellular   11   may       238         2    361         5  failure   no  
4486    unknown    9   may       514         1     -1         0  unknown   no  
4499   cellular   21   nov       362         1     -1         0  unknown   no  
4503   cellular   29   jul       816         6     -1         0  unknown  yes  

[678 rows x 17 columns]

In [17]:
secondaryDf

age          job  marital  education default  balance housing loan  \
1      33     services  married  secondary      no     4789     yes  yes   
4      59  blue-collar  married  secondary      no        0     yes   no   
7      39   technician  married  secondary      no      147     yes   no   
10     39     services  married  secondary      no     9374     yes   no   
11     43       admin.  married  secondary      no      264     yes   no   
...   ...          ...      ...        ...     ...      ...     ...  ...   
4514   38  blue-collar  married  secondary      no     1205     yes   no   
4515   32     services   single  secondary      no      473     yes   no   
4516   33     services  married  secondary      no     -333     yes   no   
4518   57   technician  married  secondary      no      295      no   no   
4519   28  blue-collar  married  secondary      no     1137      no   no   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
1     cellular   11   may       220         1    339         4  failure  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
7     cellular    6   may       151         2     -1         0  unknown  no  
10     unknown   20   may       273         1     -1         0  unknown  no  
11    cellular   17   apr       113         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4514  cellular   20   apr        45         4    153         1  failure  no  
4515  cellular    7   jul       624         5     -1         0  unknown  no  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  

[2306 rows x 17 columns]

In [19]:
tertiaryDf

age            job   marital education default  balance housing loan  \
2      35     management    single  tertiary      no     1350     yes   no   
3      30     management   married  tertiary      no     1476     yes  yes   
5      35     management    single  tertiary      no      747      no   no   
6      36  self-employed   married  tertiary      no      307     yes   no   
8      41   entrepreneur   married  tertiary      no      221     yes   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4501   34     management   married  tertiary      no      297     yes   no   
4506   42     unemployed  divorced  tertiary      no     -166      no   no   
4509   51     technician   married  tertiary      no     2506      no   no   
4517   57  self-employed   married  tertiary     yes    -3313     yes  yes   
4520   44   entrepreneur    single  tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
5     cellular   23   feb       141         2    176         3  failure  no  
6     cellular   14   may       341         1    330         2    other  no  
8      unknown   14   may        57         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4501  cellular   26   aug        63         4     -1         0  unknown  no  
4506  cellular   29   aug        85         4     -1         0  unknown  no  
4509  cellular   30   nov       210         3     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[1350 rows x 17 columns]

In [20]:
unknownDf

age          job   marital education default  balance housing loan  \
20     38   management  divorced   unknown      no        0     yes   no   
27     67      retired   married   unknown      no      696      no   no   
49     61       admin.   married   unknown      no     4629     yes   no   
132    43  blue-collar   married   unknown     yes     -715     yes   no   
133    48       admin.   married   unknown      no        0     yes   no   
...   ...          ...       ...       ...     ...      ...     ...  ...   
4337   73      retired   married   unknown      no      519      no   no   
4342   50  blue-collar   married   unknown      no     2284     yes   no   
4428   48  blue-collar   married   unknown      no      439     yes   no   
4471   59   management   married   unknown      no     3534      no   no   
4508   42       admin.   married   unknown      no      642     yes  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
20     cellular   18   nov        96         2     -1         0  unknown   no  
27    telephone   17   aug       119         1    105         2  failure   no  
49     cellular   27   jan       181         1     92         1  success  yes  
132     unknown   30   may        67         3     -1         0  unknown   no  
133    cellular    8   may        85         1    168         2  failure   no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4337  telephone   16   oct       434         1     57         1  failure  yes  
4342  telephone   31   jul      1088        17     -1         0  unknown  yes  
4428   cellular    2   feb       212         1    249         2    other   no  
4471   cellular   21   nov       216         4     -1         0  unknown   no  
4508    unknown   16   may       509         2     -1         0  unknown   no  

[187 rows x 17 columns]

#### know the unique values using unique

In [154]:
df['education'].unique()

array(['primary', 'secondary', 'tertiary', 'unknown'], dtype=object)

In [158]:
education_dfs=[]
for i in df['education'].unique():
    education_dfs.append( df[df['education']==i]  )
#     print(df[df['education']==i])
#     print(df['education']==i)

In [159]:
education_dfs[0]

age            job   marital education default  balance housing loan  \
0      30     unemployed   married   primary      no     1787      no   no   
9      43       services   married   primary      no      -88     yes  yes   
18     25    blue-collar    single   primary      no     -221     yes   no   
26     55    blue-collar   married   primary      no      627     yes   no   
36     78        retired  divorced   primary      no      229      no   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4480   23    blue-collar   married   primary      no     1158     yes   no   
4485   53    blue-collar   married   primary      no      238     yes   no   
4486   37    blue-collar   married   primary      no      378     yes   no   
4499   45    blue-collar  divorced   primary      no      942      no   no   
4503   60  self-employed   married   primary      no      362      no  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
0      cellular   19   oct        79         1     -1         0  unknown   no  
9      cellular   17   apr       313         1    147         2  failure   no  
18      unknown   23   may       250         1     -1         0  unknown   no  
26      unknown    5   may       247         1     -1         0  unknown   no  
36    telephone   22   oct        97         1     -1         0  unknown  yes  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4480   cellular   16   apr       743         1     -1         0  unknown   no  
4485   cellular   11   may       238         2    361         5  failure   no  
4486    unknown    9   may       514         1     -1         0  unknown   no  
4499   cellular   21   nov       362         1     -1         0  unknown   no  
4503   cellular   29   jul       816         6     -1         0  unknown  yes  

[678 rows x 17 columns]

In [160]:
education_dfs[1]

age          job  marital  education default  balance housing loan  \
1      33     services  married  secondary      no     4789     yes  yes   
4      59  blue-collar  married  secondary      no        0     yes   no   
7      39   technician  married  secondary      no      147     yes   no   
10     39     services  married  secondary      no     9374     yes   no   
11     43       admin.  married  secondary      no      264     yes   no   
...   ...          ...      ...        ...     ...      ...     ...  ...   
4514   38  blue-collar  married  secondary      no     1205     yes   no   
4515   32     services   single  secondary      no      473     yes   no   
4516   33     services  married  secondary      no     -333     yes   no   
4518   57   technician  married  secondary      no      295      no   no   
4519   28  blue-collar  married  secondary      no     1137      no   no   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
1     cellular   11   may       220         1    339         4  failure  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
7     cellular    6   may       151         2     -1         0  unknown  no  
10     unknown   20   may       273         1     -1         0  unknown  no  
11    cellular   17   apr       113         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4514  cellular   20   apr        45         4    153         1  failure  no  
4515  cellular    7   jul       624         5     -1         0  unknown  no  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  

[2306 rows x 17 columns]

In [161]:
education_dfs[2]

age            job   marital education default  balance housing loan  \
2      35     management    single  tertiary      no     1350     yes   no   
3      30     management   married  tertiary      no     1476     yes  yes   
5      35     management    single  tertiary      no      747      no   no   
6      36  self-employed   married  tertiary      no      307     yes   no   
8      41   entrepreneur   married  tertiary      no      221     yes   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4501   34     management   married  tertiary      no      297     yes   no   
4506   42     unemployed  divorced  tertiary      no     -166      no   no   
4509   51     technician   married  tertiary      no     2506      no   no   
4517   57  self-employed   married  tertiary     yes    -3313     yes  yes   
4520   44   entrepreneur    single  tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
5     cellular   23   feb       141         2    176         3  failure  no  
6     cellular   14   may       341         1    330         2    other  no  
8      unknown   14   may        57         2     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4501  cellular   26   aug        63         4     -1         0  unknown  no  
4506  cellular   29   aug        85         4     -1         0  unknown  no  
4509  cellular   30   nov       210         3     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[1350 rows x 17 columns]

In [162]:
education_dfs[3]

age          job   marital education default  balance housing loan  \
20     38   management  divorced   unknown      no        0     yes   no   
27     67      retired   married   unknown      no      696      no   no   
49     61       admin.   married   unknown      no     4629     yes   no   
132    43  blue-collar   married   unknown     yes     -715     yes   no   
133    48       admin.   married   unknown      no        0     yes   no   
...   ...          ...       ...       ...     ...      ...     ...  ...   
4337   73      retired   married   unknown      no      519      no   no   
4342   50  blue-collar   married   unknown      no     2284     yes   no   
4428   48  blue-collar   married   unknown      no      439     yes   no   
4471   59   management   married   unknown      no     3534      no   no   
4508   42       admin.   married   unknown      no      642     yes  yes   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
20     cellular   18   nov        96         2     -1         0  unknown   no  
27    telephone   17   aug       119         1    105         2  failure   no  
49     cellular   27   jan       181         1     92         1  success  yes  
132     unknown   30   may        67         3     -1         0  unknown   no  
133    cellular    8   may        85         1    168         2  failure   no  
...         ...  ...   ...       ...       ...    ...       ...      ...  ...  
4337  telephone   16   oct       434         1     57         1  failure  yes  
4342  telephone   31   jul      1088        17     -1         0  unknown  yes  
4428   cellular    2   feb       212         1    249         2    other   no  
4471   cellular   21   nov       216         4     -1         0  unknown   no  
4508    unknown   16   may       509         2     -1         0  unknown   no  

[187 rows x 17 columns]